### PR TITLE
DX-23332: Add method to parse date from null-terminated string

### PR DESCRIPTION
**Pull request by projjal**

### CODE REVIEW

File: cpp/src/arrow/util/value_parsing.h [hunk 1]
```diff
@@ -561,6 +562,38 @@ static inline bool ParseTimestampStrptime(const char* buf, size_t length,
   return true;
 }
 
+/// \brief Returns time since the UNIX epoch in the requested unit. Takes null terminated
+/// buffer as argument
+static inline bool ParseTimestampStrptime(const char* buf,
```
Comment (projjal) [inline cpp/src/arrow/util/value_parsing.h:567]: Line 533 contains the original function that takes non null-terminated string as argument.

File: cpp/src/arrow/util/value_parsing.h [hunk 2]
```diff
@@ -23,6 +23,7 @@
 #include <chrono>
 #include <cstddef>
 #include <cstdint>
+#include <cstring>
```
Comment (pprudhvi) [inline cpp/src/arrow/util/value_parsing.h:26]: this is needed?

**Comment by pprudhvi** [comment]
LGTM

File: cpp/src/arrow/util/value_parsing.h (hunk 2 above)
Comment (projjal) [reply cpp/src/arrow/util/value_parsing.h:26]: for the strlen function in L582

File: cpp/src/arrow/util/value_parsing.h [hunk 3]
```diff
@@ -561,6 +562,38 @@ static inline bool ParseTimestampStrptime(const char* buf, size_t length,
...
+                                          bool ignore_time_in_day,
+                                          bool allow_trailing_chars, TimeUnit::type unit,
+                                          int64_t* out) {
+  // NOTE: strptime() is more than 10x faster than arrow_vendored::date::parse().
+  struct tm result;
+  memset(&result, 0, sizeof(struct tm));
+#ifdef _WIN32
+  char* ret = arrow_strptime(buf, format, &result);
+#else
+  char* ret = strptime(buf, format, &result);
```
Comment (pravindra) [inline cpp/src/arrow/util/value_parsing.h:577]: can we put a switch here to do the copy only if buf[len-1] = '\0' ? That'll avoid duplication of code.

File: cpp/src/arrow/util/value_parsing.h [hunk 4]
```diff
@@ -561,6 +562,38 @@ static inline bool ParseTimestampStrptime(const char* buf, size_t length,
...
+#endif
+  if (ret == NULLPTR) {
+    return false;
+  }
+  if (!allow_trailing_chars && static_cast<size_t>(ret - buf) != strlen(buf)) {
```
Comment (pravindra) [inline cpp/src/arrow/util/value_parsing.h:582]: we can save the strlen if you pass length as argument.

File: cpp/src/arrow/util/value_parsing.h (hunk 3 above)
Comment (projjal) [reply cpp/src/arrow/util/value_parsing.h:577]: There is still a copy in the constructor of clean_copy. I tried running the benchmark with this but the execution time doesn't improve much if we don't remove the constructor.

Comment (projjal) [reply cpp/src/arrow/util/value_parsing.h:577]: Also c++ stl string permits '\0' as character, so we can't reject it based on that if method doesn't take null terminated string by design

File: cpp/src/arrow/util/value_parsing.h (hunk 4 above)
Comment (projjal) [reply cpp/src/arrow/util/value_parsing.h]: done

File: cpp/src/arrow/util/value_parsing.h (hunk 3 above)
Comment (pravindra) [reply cpp/src/arrow/util/value_parsing.h:577]: sounds good, you are right.

**Pull request merged**